### ISSUE TITLE
Menu remains open after actions which moves focus

### ISSUE THREAD

**Issue by karlgroves** [WordPress/gutenberg#15313]
Menu remains open after actions which moves focus

-   **Severity:** Medium
-   **Affected Populations:**
    -   Low-Vision
    -   Motor Impaired
    -   Cognitively Impaired
-   **Platform(s):**
    -   All / Universal
-   **Components affected:**
    -   Block Options

#### Issue description

When the Unified Toolbar is being used, the Options menu remains open
after triggering an action that causes the keyboard focus to be moved to
the edit area. This results in a situation where the focus position
might be underneath the menu. The actions of duplicating a block,
inserting new content before or after a block, or removing a block, all
have this behaviour.

Obscuring the focus position is potentially confusing for sighted
keyboard users, since it may appear as though no item has focus, or make
it unclear as to where the focus position is. Additionally it is simply
difficult to view the content and moving keyboard focus to and from the
toolbar is cumbersome.

#### Remediation Guidance

Whenever an action in the Options menu causes the focus to move, the
menu should close automatically.

#### Relevant standards

-   [2.1.1 Keyboard
    (Level A)](https://www.w3.org/TR/WCAG20/#keyboard-operation-keyboard-operable)
-   [2.4.7 Focus Visible (Level
    AA)](https://www.w3.org/TR/WCAG20/#navigation-mechanisms-focus-visible)

**Note:** This issue may be a duplicate with other existing accessibility-related bugs in this project. This issue comes from the Gutenberg accessibility audit, performed by [Tenon](https://www.tenon.io) and funded by [WP Campus](https://wpcampus.org/). This issue is GUT-101 in Tenon's report

**Comment by melchoyce** [comment]
Screenshot from the full report:

![image](https://user-images.githubusercontent.com/2846578/57182754-d9651000-6e57-11e9-94a9-9ee17a57d97c.png)

**Comment by afercia** [comment]
Noting this is also inconsistent depending on the state of the "Top Toolbar" option, not sure what the argumentation behind this different behavior is:

- disabled: toolbars appear on the blocks: activating one of those command in the More options menu moves the focus and closes the menu
- enabled: toolbar unified at the top: the same actions move the focus but keep the menu open

**Comment by mapk** [comment]
The solution suggested in the issue sounds like the right way to proceed. 

> Whenever an action in the Options menu causes the focus to move, the menu should close automatically.

**Comment by aduth** [comment]
This technically a duplicate of #14754.

That being said, I think the titling and context included in the comments here prove more useful toward a solution. I'll close the other in favor of this.

Including some specific comment text from #14754 in the debugging I had performed there:

At https://github.com/WordPress/gutenberg/issues/14754#issuecomment-479009098 :

> I've reproduced the issue. I assume it's likely related to how we forcibly re-insert a default block when all other blocks are removed. This re-insertion likely triggers focus to be moved to the block and out of the menu.
> 
> https://github.com/WordPress/gutenberg/blob/9af3f420d2e4d620f9a8dac8f092b8447c3d4576/packages/block-editor/src/store/actions.js#L404-L406
> 
> https://github.com/WordPress/gutenberg/blob/9af3f420d2e4d620f9a8dac8f092b8447c3d4576/packages/block-editor/src/components/block-list/block.js#L82-L86
> 
> I assume that if this analysis is correct, the bug will have been present for a while. That said, the `insertDefaultBlock` behavior was recently refactored as part of #14189 (cc @jorgefilipecosta ).
> 
> I'd suggest action items:
> 
> * Find a way to avoid focus changes as part of the "ensure default block" behavior, or reimplement the behavior such that backspacing in the first empty block will simply behave as a noop (rather than having a new block be reinstated after the fact)
> * Make sure that dropdown menus become closed when focus leaves

At https://github.com/WordPress/gutenberg/issues/14754#issuecomment-480268587 :

> Part of the problem is also that selecting the menu items does not close the menu. I'm unsure whether it was done so intentionally, but my reading of the recommendations is that the menu [should be closed](https://www.w3.org/TR/wai-aria-practices/#menu).
> 
> I wonder also if it's something we can improve in the abstractions and relationships between [`Dropdown`](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/dropdown) and [`MenuItem`](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/menu-item) components. There's also [`DropdownMenu`](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/dropdown-menu) which... is neither used by the block settings menu, nor itself uses the `MenuItem` component. At least in how it _is_ implemented, it may serve as a suitable replacement for the block settings menu's use of the more generic `Dropdown` component, with the auto-close behavior built in.
